### PR TITLE
!!! Migrate to doctrine/migration v3

### LINKED ISSUE #12: Check compatibility to Doctrine Migrations 3.x
We should evaluate if Doctrine Migrations 3.x can be used (maybe with more recent TYPO3 versions than 9.x that we originally used for developing most of this ext).

### PR DESCRIPTION
**Pull request by baschny**
This completes the open PR https://github.com/andreaswolf/typo3-ext-migrations/pull/20.

Makes the migrations work again with doctrine/migration v3. See also https://github.com/doctrine/migrations/blob/3.6.x/UPGRADE.md

### Changes
* Implement a DependencyFactory for TYPO3
* Get rid of DoctrineService
* Get rid of overwriting the `*Command` classes from doctrine/migrate which are now `final`
* Wire the dependencies in `Services.yaml`
* Get rid of obsolete `Commands.php`
* Add new commands introduced with doctrine/migrations v3:
  - migrations:dump-schema
  - migrations:generate
  - migrations:latest
  - migrations:list
  - migrations:up-to-date

### Compatibility Notes:
* `--dry-run` is currently not supported anymore for the "AbstractDataHandlerMigration"
* An upgrade in the doctrine_migrationstatus is required, see README.md

Resolves #12

### CODE REVIEW

**Comment by baschny** [comment]
@kaystrobach @andreaswolf just FYI, we have been using this branch for some time now in production with success.

**Comment by andreaswolf** [comment]
@baschny I took your updates and integrated them into the earlier PR #20.

File: README.md [hunk 1]
```diff
@@ -10,28 +12,18 @@ To execute all pending migrations you can run:
...
     <path-to-bin>/typo3 migrations:migrate
 
-This will give you an output like this:
+# Upgrade from TYPO3 9/10 to 11/12
 
+Moving to TYPO3 11, this extensions switched from doctrine/migration 2.x to 3.x for compatibility
+with PHP 8.
+
+After running the Database Compare migrations, don't forget to run the Upgrade Wizards which will
+migrate the table `migration_statustable`:
```
Comment (andreaswolf) [inline README.md:21]: `migration_statustable` is the name of the wizard, not the table, right? This is not really understandable here IMO.

File: Classes/Typo3ConfigurationLoader.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+use TYPO3\CMS\Core\Core\Environment;
+use TYPO3\CMS\Core\Database\ConnectionPool;
+use TYPO3\CMS\Core\Log\LogManager;
+use TYPO3\CMS\Core\Package\PackageInterface;
+use TYPO3\CMS\Core\Package\PackageManager;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+/**
+ * Implements a doctrine/migration ConfigurationLoader with a TYPO3 specific flavour
+ * - Adds
```
Comment (andreaswolf) [inline Classes/Typo3ConfigurationLoader.php:19]: Seems there's something missing here, right?

File: Classes/UpgradeWizard/StatusTableUpgradeWizard.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+use TYPO3\CMS\Core\Package\PackageManager;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Install\Updates\DatabaseUpdatedPrerequisite;
+use TYPO3\CMS\Install\Updates\UpgradeWizardInterface;
+
+/**
+ * Migrates the version number in the `doctrine_migrationstatus` table
+ *
+ * Transforms version number (i.e. '20230912174700') in column "version"
+ * to the fully qualified classname as expected for doctrine/migrate v3
```
Comment (andreaswolf) [inline Classes/UpgradeWizard/StatusTableUpgradeWizard.php:17]: … for doctrine/migrations v3, right?

File: README.md [hunk 4]
```diff
@@ -10,28 +12,18 @@ To execute all pending migrations you can run:
...
-    >> Available Migrations:                               1
-    >> New Migrations:                                     1
-
- == Migration Versions
-    >> 2014-07-14 18:44:53 (20140714184453) migrations                  not migrated
+bin/typo3 upgrade:run migration_statustable
 ```
 
-This extension uses `doctrine/migrations` to migrate the database tables.
+Replace the `@namespace` setting with the namespace of your extension.
```
Comment (andreaswolf) [inline README.md:26]: Which setting? I could not find any other mention of `@namespace`.

File: README.md (hunk 1 above)
Comment (baschny) [reply README.md:21]: True. But I understand it  :)

File: Classes/Typo3ConfigurationLoader.php (hunk 2 above)
Comment (baschny) [reply Classes/Typo3ConfigurationLoader.php:19]: For sure.. If I recall correctly I changed a lots of things in this loader which I noted when I started to try to document it. And then forgot to come back to it.

By reading the code, I think this description would seem appropriate:

```
* - Configure table `doctrine_migrationstatus` as where to store our migration status
* - Adds relevant TYPO3 specific paths where to search migrations:
*   1. `var/migrations/` and
*   2. `Migrations/Mysql` paths inside all active extensions
```

File: Classes/UpgradeWizard/StatusTableUpgradeWizard.php (hunk 3 above)
Comment (baschny) [reply Classes/UpgradeWizard/StatusTableUpgradeWizard.php:17]: Right!

File: README.md (hunk 4 above)
Comment (baschny) [reply README.md:26]: This line is obsolete, it was overseen and should have been deleted when I introduced the upgrade wizard (which does this automatically). See this commit: https://github.com/andreaswolf/typo3-ext-migrations/pull/23/commits/ef0aef7ec056c8974b237643507b4e17833f18f2

**Pull request merged**